validator/client/validator.go

```diff
@@ -1212,44 +1209,66 @@ func (v *validator) ChangeHost() {
 func (v *validator) filterAndCacheActiveKeys(ctx context.Context, pubkeys [][fieldparams.BLSPubkeyLength]byte, slot primitives.Slot) ([][fieldparams.BLSPubkeyLength]byte, error) {
 	ctx, span := trace.StartSpan(ctx, "validator.filterAndCacheActiveKeys")
 	defer span.End()
-
+	isEpochStart := slots.IsEpochStart(slot)
+	validatorStatusUnpopulated := false
 	filteredKeys := make([][fieldparams.BLSPubkeyLength]byte, 0)
 	statusRequestKeys := make([][]byte, 0)
+	validatorStatuses := make([]*validatorStatus, 0)
 	for _, k := range pubkeys {
-		_, ok := v.pubkeyToValidatorIndex[k]
-		// Get validator index from RPC server if not found.
+		sta, ok := v.pubkeyToStatus[k]
 		if !ok {
-			i, ok, err := v.validatorIndex(ctx, k)
-			if err != nil {
-				return nil, err
-			}
-			if !ok { // Nothing we can do if RPC server doesn't have validator index.
-				continue
-			}
-			v.pubkeyToValidatorIndex[k] = i
+			validatorStatusUnpopulated = true
+			// skip adding if not found
+			continue
 		}
 		copiedk := k
 		statusRequestKeys = append(statusRequestKeys, copiedk[:])
+		copiedStatus := sta
+		validatorStatuses = append(validatorStatuses, copiedStatus)
 	}
-	resp, err := v.validatorClient.MultipleValidatorStatus(ctx, &ethpb.MultipleValidatorStatusRequest{
-		PublicKeys: statusRequestKeys,
-	})
-	if err != nil {
-		return nil, err
+	if isEpochStart || validatorStatusUnpopulated {
+		resp, err := v.validatorClient.MultipleValidatorStatus(ctx, &ethpb.MultipleValidatorStatusRequest{
+			PublicKeys: statusRequestKeys,
+		})
+		if err != nil {
+			return nil, err
+		}
+		if resp == nil {
+			return nil, errors.New("response is nil")
+		}
+		if len(resp.Statuses) != len(resp.PublicKeys) {
+			return nil, fmt.Errorf("expected %d pubkeys in status, received %d", len(resp.Statuses), len(resp.PublicKeys))
+		}
+		if len(resp.Statuses) != len(resp.Indices) {
+			return nil, fmt.Errorf("expected %d indices in status, received %d", len(resp.Statuses), len(resp.Indices))
+		}
+		statuses := make([]*validatorStatus, 0)
+		for i, s := range resp.Statuses {
+			sta := &validatorStatus{
+				publicKey: resp.PublicKeys[i],
+				status:    s,
+				index:     resp.Indices[i],
+			}
+			statuses = append(statuses, sta)
+			// update cache
+			v.pubkeyToStatus[bytesutil.ToBytes48(resp.PublicKeys[i])] = sta
+		}
+		validatorStatuses = statuses
```

Comment:
You `continue` in the previous loop, but anyway this will be overwritten here unconditionally

Reply:
This is only if it's on epoch start/only the validator statuses that changed in the middle of the epoch

Reply:
having the continue was the bug here